flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointedInputGate.java

```diff
@@ -171,6 +176,18 @@ else if (bufferOrEvent.getEvent().getClass() == CancelCheckpointMarker.class) {
 		}
 	}
 
+	public List<Buffer> requestInflightBuffers(long checkpointId, int channelIndex) throws IOException {
+		if (((CheckpointBarrierUnaligner) barrierHandler).hasInflightData(checkpointId, channelIndex)) {
+			return inputGate.getChannel(channelIndex).requestInflightBuffers(checkpointId);
```

Comment:
I think that `CheckpointBarrierUnaligner.hasInflightData` and the actual InflightBuffers in the input channel
can be inconsistent as they are updated by different threads.

Reply:
`CheckpointBarrierUnaligner#hasInflightData` is only updated in task thread. So while calling `requestInflightBuffers` (on gate/channel) from task thread, the state cannot change.
At this point, `receivedBuffers` of input channel can also not be drained by task thread, so we are not losing any buffer. However, new buffers could be added by netty thread causing duplicate buffers.
To guard against this case, `RemoteInputChannel#requestInflightBuffers` first acquires lock and copies the buffers. It then updates `lastRequestedCheckpointId`, so that all subsequently added `receivedBuffers` should go through `BufferReceivedListener`.

Reply:
I mean the case when:
1. task thread is suspended by OS scheduler
1. netty thread receives a buffer `X` and adds it to `RemoteInputChannel.receivedBuffers`
1. netty thread receives a barrier and schedules a checkpoint by enqueueing a mail (`RIC.onBuffer` -> `ThreadSafeUnaligner#notifyBarrierReceived`)
1. task thread wakes up and sees a **mail** (not a default action);
1. task thread performs a checkpoint; it checks `CheckpointBarrierUnaligner.hasInflightData` and reads `false` so it doesn't go into `requestInflightBuffers`

Now:
- The buffer `X` added in the beginning is missing in the checkpoint; 
- `RemoteInputChannel#lastRequestedCheckpointId` is not updated;

Am I missing something?

Reply:
When a barrier is received in the input channel, it triggers the notify, but is also added at the tail of the queue in its in-order position.
`hasInflightData` is only updated to `false` in `Unaligner#processBarrier` when task thread pulls the barrier from its in-order position. That means, that `hasInflightData` can only be false iff `X` has been consumed in the meantime.

Reply:
I see, thanks for the explanation.